data/transactions/logic/resources.go

```diff
@@ -0,0 +1,328 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// resources contains a catalog of available resources. It's used to track the
+// apps, assets, and boxes that are available to a transaction, outside the
+// direct foreign array mechanism.
+type resources struct {
+	// These resources were created previously in the group, so they can be used
+	// by later transactions.
+	createdAsas []basics.AssetIndex
+	createdApps []basics.AppIndex
```

Comment:
My suggestion for solving the problem that apps currently in the process of being created cannot be referenced by inner txns (if you're looking to solve it in this PR):
* Create a new slice/map of `basics.AssetIndex` here to track apps that are being created, but whose creation txns haven't yet finished. Let's call it `creatingApps`
* In `EvalContract`, if this is an app create, add the app ID to `creatingApps`
* In `RecordAD`, if the txn was an app create, remove the app ID from `creatingApps` and add it to `createdApps` (it already does the second part)
* When you reference `createdApps` in `resources.go` and `eval.go`, you can now also potentially consult `creatingApps`. Note that since `EvalContext.availableApp` looks at `createdApps` since `createdResourcesVersion` (6), we probably _don't_ want to consult `creatingApps` until `resourceSharingVersion`, to preserve backwards compatibility.

Since there are backwards compatibility concerns, this approach seems better than modifying `createdApps` directly, or changing when `RecordAD` gets invoked

Reply:
I decided the retain as much existing code as possible to limit the chance of a backward incompatible change. So, I simply retained the existing rule, that the app account of the current application id is a legal address. I didn't change things to make that address available to contracts that are _called_ by the app being created (except to the extent that it becomes available by being the Sender of a child call, or perhaps passed in tx.Accounts even further down.)

I did this with this test:
```
	if cx.txn.Txn.ApplicationID == 0 && cx.getApplicationAddress(cx.appID) == addr {
```
in allowsLocals

Reply:
Makes sense